### PR TITLE
feat!: single geant4 simulation algorithm (ActsExamples 1/X)

### PR DESCRIPTION
**Pull request by asalzburger**
This PR is the first of a series of PRs that will:

 * harmonise the Geant4 examples (and running setup)
 * enable Geant4 simulation with sensitive elements
 * harmonise Fatras/Geant4 simulation to run with same commands
 * streamline the single particle generation (currently done in ParticleGun, PropagationAlgorithm, Geant4 gun)

This PR addresses at first the former `GeantintoRecording`:

 * `GeantinoRecording` application is renamed to `MaterialRecording` 
 *  the material recording is now done by the `Geant4Simulation` algorithm:
     * it reads in generated particles from the `ParticleGun` and feeds into the new `PrimaryGeneratorAction`, as a consequence the specific `g4-gen-*` options are removed, as this is covered by the generator options now
     * the singleton `instance` framework for user actions is removed in favour of the new event-wise  thread-safe `EventStoreRegistry`
     * the simulation dependent internal collection names are moved into `Simulation/Common`
 *  the common code base of `MaterialMappingBase` is renamed (alongside all other Examples) to `CommonMaterialRecording` with 
 * the `MaterialInteraction` and its derivatives are removed from the `Propgator/MaterialInteractor.hpp` into a separate file `Material/MaterialInteraction.hpp` to wich the former file and the other clients point to

`BREAKING CHANGE:` 

Core:
 * the `MaterialInteraction` and its derivatives are removed from the `Propgator/MaterialInteractor.hpp` into a separate file `Material/MaterialInteraction.hpp` to wich the former file and the other clients point to
 
Examples (not yet declared as breaking):
 * all the renaming discussed above, removal of algorithms, etc.

### CODE REVIEW

File: Examples/Algorithms/Geant4/src/Geant4Simulation.cpp [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+  // Output handling: Material tracks
+  if (not m_cfg.outputMaterialTracks.empty()) {
+    ctx.eventStore.add(
+        m_cfg.outputMaterialTracks,
+        std::move(EventStoreRegistry::recordedMaterial[ctx.eventNumber]));
+  }
+
+  // Clear the EventStoreRegistry
+  EventStoreRegistry::clearEvent(ctx.eventNumber);
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Algorithms/Geant4/src/Geant4Simulation.cpp:149]: Suggestion by Paul: `std::map` and `erase()`.

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/EventStoreRegistry.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+///
+/// @note multiple threads within an event could lead to conflicts
+class EventStoreRegistry {
+ public:
+  /// Constructore with arguments
+  /// @param nevents is the number of events to be processed
+  EventStoreRegistry(size_t nevents);
+  virtual ~EventStoreRegistry() = default;
+
+  static std::vector<WhiteBoard*> boards;
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Algorithms/Geant4/include/ActsExamples/Geant4/EventStoreRegistry.hpp]: I changed to a map as suggested below.

File: Examples/Algorithms/Geant4/src/EventStoreRegistry.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+std::vector<ActsExamples::SimParticleContainer::sequence_type>
+    ActsExamples::EventStoreRegistry::particlesFinal =
+        std::vector<ActsExamples::SimParticleContainer::sequence_type>();
+
+std::vector<std::vector<Acts::RecordedMaterialTrack>>
+    ActsExamples::EventStoreRegistry::recordedMaterial =
+        std::vector<std::vector<Acts::RecordedMaterialTrack>>();
+
+ActsExamples::EventStoreRegistry::EventStoreRegistry(size_t nevents) {
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Algorithms/Geant4/src/EventStoreRegistry.cpp]: See comment above.

File: Examples/Algorithms/Geant4/src/Geant4Simulation.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    g4World->GetLogicalVolume()->SetFieldManager(fieldMgr, true);
+  }
+}
+
+ActsExamples::Geant4Simulation::~Geant4Simulation() {}
+
+ActsExamples::ProcessCode ActsExamples::Geant4Simulation::execute(
+    const ActsExamples::AlgorithmContext& ctx) const {
+  // Ensure exclusive access to the geant run manager
+  std::lock_guard<std::mutex> guard(m_runManagerLock);
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Algorithms/Geant4/src/Geant4Simulation.cpp:98]: I kept the `std::lock_guard` for the moment, but changed to a map.

File: Examples/Algorithms/Geant4/src/Geant4Simulation.cpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  // Output handling: Simulated hits
+  if (not m_cfg.outputSimHits.empty()) {
+    SimHitContainer simHits;
+    simHits.insert(EventStoreRegistry::hits[ctx.eventNumber].begin(),
+                   EventStoreRegistry::hits[ctx.eventNumber].end());
+    EventStoreRegistry::hits[ctx.eventNumber].clear();
+    // Register to the event store
+    ctx.eventStore.add(m_cfg.outputSimHits, std::move(simHits));
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Algorithms/Geant4/src/Geant4Simulation.cpp]: They are different types, what is stored in the hit production is a `SimHitContainer::sequece_type` and what's written to the EventStore is `SimHitContainer`.

**Review by Corentin-Allaire — COMMENTED** [review]
I went through the PR and it looks al good to me. I found a few typo that I put in comments.

File: Examples/Algorithms/CMakeLists.txt [hunk 6]
```diff
@@ -3,7 +3,7 @@ add_subdirectory_if(HepMC ACTS_BUILD_EXAMPLES_HEPMC3)
 add_subdirectory(Fatras)
 add_subdirectory_if(Geant4HepMC ACTS_BUILD_EXAMPLES_GEANT4 AND ACTS_BUILD_EXAMPLES_HEPMC3)
 add_subdirectory_if(Geant4 ACTS_BUILD_EXAMPLES_GEANT4)
-add_subdirectory_if(Geant4DD4hep ACTS_BUILD_EXAMPLES_DD4HEP AND ACTS_BUILD_EXAMPLES_GEANT4)
+# add_subdirectory_if(Geant4DD4hep ACTS_BUILD_EXAMPLES_DD4HEP AND ACTS_BUILD_EXAMPLES_GEANT4)
```
Comment (paulgessinger) [inline Examples/Algorithms/CMakeLists.txt:6]: Obsolete? If so can we remove this line?

File: Examples/Algorithms/Geant4/include/ActsExamples/DDG4/DDG4DetectorConstruction.hpp [hunk 7]
```diff
@@ -18,26 +18,32 @@ class Detector;
 namespace ActsExamples {
 
 /// Construct the Geant4 detector from a DD4hep description.
-class DD4hepDetectorConstruction final : public G4VUserDetectorConstruction {
+class DDG4DetectorConstruction final : public G4VUserDetectorConstruction {
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Examples/Algorithms/Geant4/include/ActsExamples/DDG4/DDG4DetectorConstruction.hpp:21]: 👍

File: docs/howto/run_material_mapping.rst [hunk 8]
```diff
@@ -54,23 +54,23 @@ Then edit the config-map.json file
...
 ----------------
 
-With the surfaces map and the material track we can finally do the material mapping using the ``MaterialMapping`` algorithm :
+With the surfaces map and the material track we can finally do the material mapping using the ``MaterialMapping`` application :
 
 .. code-block:: console
 
-  ./../build/bin/ActsExampleMaterialMappingDD4hep -j1 --input-root true --input-files geant4-material-tracks.root --mat-input-type file --mat-input-file geometry-map.json --output-root --output-json --output-cbor --mat-mapping-collection material-tracks --mat-output-file material-maps --mat-mapping-surfaces true --mat-mapping-volumes true --mat-mapping-volume-stepsize 1 --dd4hep-input ../thirdparty/OpenDataDetector/xml/OpenDataDetector.xml
+  ./../build/bin/ActsExampleMaterialMappingDD4hep -j1 --input-root true --input-files geant4_material_tracks.root --mat-input-type file --mat-input-file geometry-map.json --output-root --output-json --output-cbor --mat-mapping-collection material-tracks --mat-output-file material-maps --mat-mapping-surfaces true --mat-mapping-volumes true --mat-mapping-volume-stepsize 1 --dd4hep-input ../thirdparty/OpenDataDetector/xml/OpenDataDetector.xml
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply docs/howto/run_material_mapping.rst:73]: Yeah, I still need to check that the tests run through ...

[the remark this answers is not in the document]
Comment (asalzburger) [reply docs/howto/run_material_mapping.rst:80]: fixed.

File: Examples/Algorithms/CMakeLists.txt (hunk 6 above)
Comment (asalzburger) [reply Examples/Algorithms/CMakeLists.txt]: Removed.

File: Examples/Algorithms/Geant4/src/SimParticleTranslation.cpp [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+  if (eventStore == nullptr) {
+    ACTS_WARNING("No EventStore instance could be found for this event!");
+    return;
+  }
+
+  // Get the number of input particles
+  const auto inputParticles =
+      eventStore->get<ActsExamples::SimParticleContainer>(m_cfg.inputParticles);
+
+  // Feserve appropriate resources for initial/final particles
```
[the remark this answers is not in the document]
Comment (asalzburger) [reply Examples/Algorithms/Geant4/src/SimParticleTranslation.cpp]: Fixed.

**Comment by asalzburger** [comment]
Comments addressed.

**Review by Corentin-Allaire — APPROVED** [review]

**Comment by paulgessinger** [comment]
@asalzburger can you make sure the fix from #912 is included in this?

**Comment by paulgessinger** [comment]
I'll merge this once you confirm that #912 is integrated, ok?

**Comment by asalzburger** [comment]
I have not touched the `EventRecording.cpp` and related `HepMC` part at all in this PR, this should come in (3/X).

**Comment by asalzburger** [comment]
I think this does not have to be rebased, the Alignment-KF inclusion is practically orthogonal to that.

**Pull request merged**